p2p: transport agnostic peer abstractions (#708)
informalsystems/tendermint-rs: docs/architecture/adr-009-transport-agnostic-peer-abstraction.md

over which transport is used to connect and converse with other peers. Given
strongly tailored abstract interfaces, concrete implementations will be easy to
spin up and plug into the machinery which lifts bytes from the wire into the
core domain and transports messages into the rest of the system.

## Decision

### Transport

Wraping the design is the `Transport`. Modelled with the properties of
a physical network endpoint in mind, which can be bound and stopped. It should
strongly correspond to the acquisition and lifecycle management of network
resources on the system.

``` rust
pub trait Transport {
    type Connection: Connection;
    type Endpoint: Endpoint<Connection = <Self as Transport>::Connection>;
    type Incoming: Iterator<Item = Result<<Self as Transport>::Connection>> + Send;

    fn bind(&self, bind_info: BindInfo) -> Result<(Self::Endpoint, Self::Incoming)>;
    fn shutdown(&self) -> Result<()>;
}
```

After the successful bind the caller holds an `Endpoint` as well as a stream of
incoming `Connection`s. Which is a standardised way to connect to new peers and
react to newly connected ones respectively.

``` rust
pub trait Endpoint: Send {
    type Connection;

>     fn connect(&self, info: ConnectInfo) -> Result<Self::Connection>;

Reviewer: What goes into `ConnectInfo`? Is it transport specific?

Author: I wouldn't assume so. It would need some `node::Id`, `ip:port`/`dns endpoint` afaict.

>     fn listen_addrs(&self) -> SocketAddr;

Reviewer: ```suggestion
    fn listen_addrs(&self) -> Vec<SocketAddr>;
```

or 

```
    fn listen_addr(&self) -> SocketAddr;
```

>     fn listen_addrs(&self) -> SocketAddr;

Reviewer: ```suggestion
    fn listen_addrs(&self) -> Vec<SocketAddr>;
```

or 

```
    fn listen_addr(&self) -> SocketAddr;
```

>     fn listen_addrs(&self) -> Vec<SocketAddr>;

Reviewer: 1) can we add a comment here and for all other functions as well?

2) can we move this into `Incoming`? it makes more sense since the iterator (TcpListener.incoming() in case of MConnection) is being used there.

Author: > can we add a comment here and for all other functions as well?

Will be addressed together with all other clippy warnings, once it's final.

> can we move this into `Incoming`? it makes more sense since the iterator (TcpListener.incoming() in case of MConnection) is being used there.

Not quite getting the meaning here. `Incoming` is an iterator, not a trait that is implemented.

Reviewer: > Not quite getting the meaning here. Incoming is an iterator, not a trait that is implemented.

please ignore

}
```

Centerpiece of the whole shebang is the `Connection`. It represents a connected
peer and provides the primitives to get data and send data from a peer. It is

> designed with the outlook to support stream based trasnports down the road.

Reviewer: ```suggestion
designed with the outlook to support stream based transports down the road.
```

While being open to enable feature parity with current production installations
based on tendermint-go's `MConn`.

``` rust
pub trait Connection: Send {

>     type Error: 'static + fmt::Display + std::error::Error + Send + Sync;

Reviewer: Is it necessary for the `Error` to have a static lifetime?

Author: I'd be surprise if we can get away without it as we likely we want to pass it around for tracing/introspection. We can omit it in the ADR.

>     type Error: 'static + fmt::Display + std::error::Error + Send + Sync;

Reviewer: Is it necessary for the `Error` to have a static lifetime?

    type Read: Read;
    type Write: Write;

    fn advertised_addrs(&self) -> Vec<SocketAddr>;
    fn close(&self) -> Result<()>;
    fn local_addr(&self) -> SocketAddr;
    fn open_bidirectional(
        &self,

>         stream_id: &StreamId,

Reviewer: does this have to be a pointer?

Author: Not necessarily, we can likely have `Copy` on `StreamId`, it's just good practice if the caller doesn't consume it to just pass a reference. Does it give you trouble in the implementation?

Reviewer: I thought for simple types (int, uint, etc.) it is better to copy. It's certainly a case for Go https://twitter.com/akaliaev/status/1370274749545320448

    ) -> Result<(Self::Read, Self::Write), Self::Error>;
    fn public_key(&self) -> PublicKey;

>     fn remote_addr(&self) -> SocketAddr;

Author: For the closed case we should be able to return the information from when the connection was running. Is there ever the case where we have a connection that is not open yet?

>     fn remote_addr(&self) -> SocketAddr;

Reviewer: Right, I was just looking at `TcpStream` `local_addr` and `peer_addr` methods that both return `Result<SocketAddr>` (rather than `SocketAddr`) even though it's also not possible to construct a non-connected stream. 

At the moment, this is how `local_addr` implemented:

```rust
    fn local_addr(&self) -> SocketAddr {
        self.stream
            .local_addr()
            .unwrap()
    }
```

if there's an error, the program will panic, which is not good.

>     fn remote_addr(&self) -> SocketAddr;

Author: It doesn't need to be done that way, right? We can also memorise this information when we construct the connection and return the stored value whenever the method is called. Or am I missing an important detail here? In my understanding the address shouldn't change over the lifetime of the conn.

}
```

### Peer

Given a correct implementation of a `Transport` and its `Connection` newly
established ones will be wrapped with a `Peer`. Which is in charge of setting
up the correct streams on the `Connection` and multiplex messages - incoming
and outgoing alike - efficiently. It's also an attempt to enforce
correct-by-construction constraints on the state machine of the peer. To avoid
misuse or unexpected transitions. The only way to construct is, is from an
existing connection which gives the caller a connected peer. When invoking run
on that one a fully function peer is "returned". Therefore the states look
like: `Connected -> Running -> Stopped`.

``` rust
impl<Conn> Peer<Connected<Conn>>
where
    Conn: Connection,
{
    pub fn run(self, stream_ids: Vec<StreamId>) -> Result<Peer<Running<Conn>>> {
        // ...
    }

>     fn stop(self) -> Result<Peer<Stopped>> {

Reviewer: why it's possible to stop `Connected` (not `Running`) peer?

Author: Both `Connected` and `Running` have an associated `stop` method. :)

Author: I forgot to include it in the ADR, the code has it.

Author: Actually it's in there. See above.

Reviewer: > Both Connected and Running have an associated stop method. :)

yeah, so my question is why it's possible to stop a non-running peer? it seems weird to me

Author: Ah right, I miss-read your question. From my perspective a `Connected` peer already holds onto some resources (i.e. the connection)  and maybe there is a valid use-case where it shouldn't be transitioned into running state but dropped before.
While we can rely on the `Drop` semantics alone I wanted a transition in the terminal `Stopped` state for both "starting" states.

        // ...
    }
}

impl<Conn> Peer<Running<Conn>>
where
    Conn: Connection,
{
    pub fn send(&self, message: message::Send) -> Result<()> {

>         // ...

Reviewer: do I understand correctly that routing (mapping message::Send type onto a certain stream) happens here?

Author: That's exactly right, the message multiplexing over the streams is responsibility of the peer implementation. Maintaining this across transports seems an important guarantee to provide.

of incoming messages can be achieved by draining the `Receiver` part of the
running state.

### Supervisor

The `Supervisor` is the main entry point to the p2p package giving higher-level
components access to a unified stream of peer events and messages as well as
the ability to control peer lifecycle (connect, disconnect, etc.).

``` rust
pub enum Command {
    Accept,
    Connect(SocketAddr),
    Disconnect(node::Id),
    Msg(node::Id, message::Send),
}

pub enum Event {
    Connected(node::Id, Direction),
    Disconnected(node::Id, Report),
    Message(node::Id, message::Receive),
    Upgraded(node::Id),
    UpgradeFailed(node::Id, Report),
}

impl Supervisor {
    pub fn run<T>(transport: T) -> Result<Self>
    where
        T: transport::Transport + Send + 'static,
    {
        // ...
    }

    pub fn recv(&self) -> Result<Event> {
        // ...
    }

    pub fn command(&self, cmd: Command) -> Result<()> {
        // ...
    }
}
```

## Status

Proposed

## Consequences

### Positive

* Unified way to bootstrap and integrate transports
* Potential for connecting different wire transports in the same process

> * Rest of the domain is simply concerned with `PeerId`s as identity

Reviewer: ```suggestion
* Rest of the domain is simply concerned with `node::Id`s as identity
```

### Negative

* Significant set of abstractions need to be satisfied for a new transport
  implementation
* Non-stream based transports need to be fitted into this model

### Neutral

## Open Questions

> * Should serialization be transport specific to allow for optimisations on the

Reviewer: NOTE: tendermint-go has decided on unified serialization (protobuf) across all transports for the sake of simplicity.

Author: Great, thanks for the context. If this is indeed the vision @brapse @greg-szabo @erikgrinaker? We should stick to the plan that connections provide byte streams and we maintain a single implementation of the serialisation.

Personally I would benefit from a brush up on how currently the byte layout looks on the wire with TCP connections. Are we still byte-prefixing to determine the `channel`?

learnings of the shortcomings of the original Go implementation can be used to
prevent certain mistakes. Namely two:

* Leakage of physical concerns into the core domain
* Flexibility to adopt different wire protocols for transport of messages

For that, the first set of newly introduced concepts will attempt to be generic
over which transport is used to connect and converse with other peers. Given
strongly tailored abstract interfaces, concrete implementations will be easy to
spin up and plug into the machinery which lifts bytes from the wire into the
core domain and transports messages into the rest of the system.

## Decision

### Transport

Wraping the design is the `Transport`. Modelled with the properties of
a physical network endpoint in mind, which can be bound and stopped. It should
strongly correspond to the acquisition and lifecycle management of network
resources on the system.

``` rust
pub trait Transport {
    type Connection: Connection;
    type Endpoint: Endpoint<Connection = <Self as Transport>::Connection>;
    type Incoming: Iterator<Item = Result<<Self as Transport>::Connection>> + Send;

>     fn bind(&self, bind_info: BindInfo) -> Result<(Self::Endpoint, Self::Incoming)>;

Author: Indeed shutdown is pointless as the transport itself is consumed. Shutdown semantics should happen when the endpoint is dropped. Updated the ADR and the code.

>     fn shutdown(&self) -> Result<()>;

Author: https://github.com/informalsystems/tendermint-rs/pull/708#discussion_r648122718

}
```

After the successful bind the caller holds an `Endpoint` as well as a stream of
incoming `Connection`s. Which is a standardised way to connect to new peers and
react to newly connected ones respectively.

``` rust
pub trait Endpoint: Send {
    type Connection;

    fn connect(&self, info: ConnectInfo) -> Result<Self::Connection>;
    fn listen_addrs(&self) -> Vec<SocketAddr>;
}
```

Centerpiece of the whole shebang is the `Connection`. It represents a connected
peer and provides the primitives to get data and send data from a peer. It is
designed with the outlook to support stream based transports down the road.
While being open to enable feature parity with current production installations
based on tendermint-go's `MConn`.

``` rust
pub trait StreamSend {

>     fn send(msg: Vec<u8>) -> Result<()>;

Author: Until we see strong reasons to go with `Bytes` the generic approach will suffice. Updated ADR and code.
